crates/turborepo-lib/src/rewrite_json.rs

```diff
@@ -0,0 +1,411 @@
+use jsonc_parser::{errors::ParseError, parse_to_ast};
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum RewriteError {
+    #[error("The JSON document failed to parse. {0}")]
+    ParseError(ParseError),
+    #[error("The JSON document contains no root object.")]
+    NoRoot,
+}
+/**
+ * When generating replacement content it's one of two things.
+ */
+enum GenerateType {
+    Object,
+    Path,
+}
+
+/**
+ * Range stores the primitive AST token details required for document
+ * mutation.
+ */
+struct Range {
+    pub start: usize,
+    pub end: usize,
+    pub replacement_char: String,
+}
+
+/**
+ * Given a JSONC document, an object traversal path, and a _pre-serialized_
+ * JSON value `set_path` will return a minimally-mutated JSONC document with
+ * the path specified set to the JSON value.
+ *
+ * - If the path exists, it will clobber the existing contents.
+ * - If the path does not exist, it will synthesize object members and
+ *   objects to ensure it does exist.
+ *
+ * In the event that the key appears multiple times in the document the last
+ * instance of the key will be updated and earlier instances will be
+ * disregarded.
+ */
+pub fn set_path(
+    json_document_string: &str,
+    path: Vec<&str>,
+    json_value: &str,
+) -> Result<String, RewriteError> {
+    let root = get_root(json_document_string)?;
+
+    // Find the token we'll be modifying and its path from the root.
+    let (closest_path, closest_node) = get_closest_node(&root, &path, vec![]);
+
+    // Pull the token metadata off of the token.
+    let (property_count, range): (usize, jsonc_parser::common::Range) = match closest_node {
+        jsonc_parser::ast::Value::Object(literal) => (literal.properties.len(), literal.range),
+        jsonc_parser::ast::Value::StringLit(literal) => (0, literal.range),
+        jsonc_parser::ast::Value::NumberLit(literal) => (0, literal.range),
+        jsonc_parser::ast::Value::BooleanLit(literal) => (0, literal.range),
+        jsonc_parser::ast::Value::Array(literal) => (0, literal.range),
+        jsonc_parser::ast::Value::NullKeyword(literal) => (0, literal.range),
+    };
+
+    // Figure out what we should be generating:
+    // - An object to be assigned to an existing member. ("object")
```

Comment:
I'm not sure these names are really clear. Could it be called `Append` or `Mutate`?

Reply:
I was trying to choose names based upon the JSON spec and be consistent to that, but I didn't do a great job.

- `{ "foo": "bar" }`: object
- `"foo": value`: member where value is any valid, pre-serialized JSON thing.

`jsonc_parser` calls `members` `properties`, which I also used when adjacent to objects they provide. 🤷‍♂️